src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -92,162 +98,346 @@ fn default_endpoint() -> String {
+                            reason: "Error when sending timeseries to GCP",
+                        });
+                    }
+
+                    break;
+                }
+
+                last_time = Instant::now();
             }
-            MetricValue::Gauge { value } => {
-                let interval = gcp::GcpInterval {
-                    start_time: None,
-                    end_time,
-                };
 
-                (*value, interval, gcp::GcpMetricKind::Gauge)
+            tokio::select! {
+                _ = stopwatch.tick() => {}
+
+                events = input.next() => {
+                    if events.is_none() {
+                        return Ok(());
+                    }
+
+                    if let EventArray::Metrics(metrics) = events.unwrap() {
+                        let start_time = time_tracker.start_time();
+                        buffer.extend(metrics.into_iter().filter_map(|m| MetricDataPoint::from_metric(&self.config, start_time, m)));
+                    }
+                }
             }
-            _ => unreachable!(),
-        };
+        }
+    }
+}
 
-        let metric_labels = series
-            .tags
-            .unwrap_or_default()
-            .into_iter_single()
-            .collect::<std::collections::HashMap<_, _>>();
-
-        let series = gcp::GcpSeries {
-            time_series: &[gcp::GcpSerie {
-                metric: gcp::GcpMetric {
-                    r#type: metric_type,
-                    labels: metric_labels,
-                },
-                resource: gcp::GcpResource {
-                    r#type: self.config.resource.r#type.clone(),
-                    labels: self.config.resource.labels.clone(),
-                },
-                metric_kind,
-                value_type: gcp::GcpValueType::Int64,
-                points: &[gcp::GcpPoint {
-                    interval,
-                    value: gcp::GcpPointValue {
-                        int64_value: Some(point_value as i64),
-                    },
-                }],
-            }],
-        };
+/// The end time must not be earlier than the start time, and the end time must not be more than 25 hours
+/// in the past or more than five minutes in the future.
+/// https://cloud.google.com/monitoring/api/ref_v3/rpc/google.monitoring.v3#timeinterval
+const GCP_25_HOURS_TIME_WINDOW: Duration = Duration::from_secs(25 * 3_600);
+
+/// Keeps track of the start time that would be use when sending time series to GCP.
+struct StartTimeTracker {
+    time: DateTime<Utc>,
+    clock: Instant,
+}
 
-        let body = crate::serde::json::to_bytes(&series).unwrap().freeze();
+impl StartTimeTracker {
+    fn new() -> Self {
+        Self {
+            time: Utc::now(),
+            clock: Instant::now(),
+        }
+    }
+
+    /// Returns the current window starting time.
+    fn start_time(&mut self) -> DateTime<Utc> {
+        if self.clock.elapsed() >= GCP_25_HOURS_TIME_WINDOW {
```

Comment:
I think the `DateTime` type should support an elapsed type operation as well to avoid needing to track two times.